lms/services/h_operations.py

```diff
@@ -0,0 +1,33 @@
+from lms.services import HAPIError
+from lms.values import HUser
+
+
+__all__ = ["HypothesisOperationsService"]
+
+
+class HypothesisOperationsService:
+    """
+    High-level h API service.
+
+    The ``hapi`` service provides a low-level client for the Hypothesis "h"
+    API. This service provides higher-level operations (eg. fetching a user)
+    on top of that client.
+    """
+
+    def __init__(self, _context, request):
+        self.hapi_svc = request.find_service(name="hapi")
+        self.request = request
+
+    def fetch_user(self, username):
+        authority = self.request.registry.settings["h_authority"]
+        userid = f"acct:{username}@{authority}"
+
+        try:
+            user_info = self.hapi_svc.get(path=f"users/{userid}").json()
+            return HUser(
+                authority=authority,
+                username=username,
+                display_name=user_info["display_name"],
+            )
+        except HAPIError as err:
+            raise err
```

Comment:
At present there is no wrapping of the exception type from the lower-level code here. This just makes it explicit that we're choosing to let these errors propagate. I'm not sure there is a whole lot of value in wrapping the exception at this level. Thoughts?

Reply:
`HAPIError` is already a services error (its full name is `lms.services.HAPIError`) so I agree there's no need to wrap it in a different exception class.

I think it's good to be explicit about re-raising in the code here. I'd also add a docstring, and document this in the docstring with a `:raise`. The reason is that what exceptions it throws and when/why is an important part of the interface of a service, and is something that someone trying to write a view that uses the service needs to know (and I wouldn't want them to have to dig in to the lower-level `hapi_svc` to spot it).

On the other hand, when _views_ call services that might raise exceptions, I have _not_ had the views explicitly catch and re-raise exceptions like this. They just let them fly. (And the exceptions are later caught be error views.) Given that we don't write code that calls our view code, I think in the case of views it's nicer to keep the code cleaner.